staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go

```diff
@@ -21,39 +21,254 @@ import (
+// namespaces.
+func UseHostNamespaces(p *corev1.Pod) {
+	p.Spec.HostNetwork = true
+	p.Spec.HostPID = true
+	p.Spec.HostIPC = true
+}
+
+// ShareProcessNamespace configures all containers in the pod to share the
+// process namespace.
+func ShareProcessNamespace(p *corev1.Pod) {
+	p.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+}
+
+// ClearSecurityContext clears the security context for the container.
+func ClearSecurityContext(p *corev1.Pod, containerName string, mask ContainerType) {
+	modifyContainer(p, containerName, mask, func(c *corev1.Container) {
+		c.SecurityContext = nil
+	})
+}
+
+// DisallowRoot configures the container to run as a non-root user.
+func DisallowRoot(p *corev1.Pod, containerName string, mask ContainerType) {
+	modifyContainer(p, containerName, mask, func(c *corev1.Container) {
+		c.SecurityContext = &corev1.SecurityContext{
+			RunAsNonRoot: pointer.BoolPtr(true),
+		}
+	})
+}
+
+// DropCapabilities drops all Capabilities for the container
+func DropCapabilities(p *corev1.Pod, containerName string, mask ContainerType) {
+	modifyContainer(p, containerName, mask, func(c *corev1.Container) {
+		if c.SecurityContext == nil {
+			c.SecurityContext = &corev1.SecurityContext{
+				Capabilities: &corev1.Capabilities{
+					Drop: []corev1.Capability{"ALL"},
+				},
+			}
+			return
+		}
+		if c.SecurityContext.Capabilities == nil {
+			c.SecurityContext.Capabilities = &corev1.Capabilities{
+				Drop: []corev1.Capability{"ALL"},
 			}
-			container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
-				MountPath: "/host",
-				Name:      volumeName,
-			})
+			return
 		}
+		c.SecurityContext.Capabilities.Drop = append(c.SecurityContext.Capabilities.Drop, "ALL")
+	})
+}
 
-		pod.Spec.HostIPC = true
-		pod.Spec.HostNetwork = true
-		pod.Spec.HostPID = true
-		return nil
-	default:
-		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
+// AllowProcessTracing grants the SYS_PTRACE capability to the container.
+func AllowProcessTracing(p *corev1.Pod, containerName string, mask ContainerType) {
+	modifyContainer(p, containerName, mask, func(c *corev1.Container) {
+		if c.SecurityContext == nil {
+			c.SecurityContext = &corev1.SecurityContext{
+				Capabilities: &corev1.Capabilities{
+					Add: []corev1.Capability{"SYS_PTRACE"},
+				},
+			}
+			return
+		}
+		if c.SecurityContext.Capabilities == nil {
+			c.SecurityContext.Capabilities = &corev1.Capabilities{
+				Drop: []corev1.Capability{"SYS_PTRACE"},
+			}
+			return
+		}
+		c.SecurityContext.Capabilities.Drop = append(c.SecurityContext.Capabilities.Drop, "SYS_PTRACE")
+	})
+}
+
+// modifyContainer performs the modifier function m against a container from cs
+// which has the name of containerName and the type satisfying the mask.
+func modifyContainer(pod *corev1.Pod, containerName string, mask ContainerType, m func(*corev1.Container)) {
```

Comment:
I see that this function does what is intended, but it looks to me too clever than simpler.

Reply:
Yeah, the bit mask is tricky, but I took the idea from [here](https://github.com/kubernetes/kubernetes/blob/6ce055d62d8608842db18901c948c6982c74a164/pkg/api/pod/util.go#L61) 😀

I think most programmers should be familiar with bit masks because the technique is prevalent in the construction of linux syscall arguments as well as certain places in the go std lib, [io/fs.FileMode](https://pkg.go.dev/io/fs#FileMode) for example.

Reply:
I don't think we can import that dependency based on kubectl's [Contribution Requirement](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubectl#contribution-requirements).

The `VisitContainers` utility function works with init containers, which we don't currently need. Since what we have satisfies our needs already and what we have is currently unexported, I think we should keep things as they are.

Reply:
Asking about this on slack, @ardaguclu pointed out that there's already a [podutils](https://github.com/kubernetes/kubernetes/blob/43e8ebbbcd3f57d18d8151efb6242f88a763b06d/staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go#L60) and suggested copying the `VisitContainers` implementation there.

Do you mind doing the copy and updating profiles to use `VisitContainers`?

Ref: https://kubernetes.slack.com/archives/C2GL57FJ4/p1675803799358839

Reply:
Updated in 5db0224fac026f01fcd5a8ffe166da04b6b5262e